File: university-service/src/main/java/ua/com/foxminded/courseproject/service/StudentServiceImpl.java
package ua.com.foxminded.courseproject.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import ua.com.foxminded.courseproject.dto.StudentDto;
import ua.com.foxminded.courseproject.exceptions.StudentNotFoundException;
import ua.com.foxminded.courseproject.mapper.StudentMapper;
import ua.com.foxminded.courseproject.repository.StudentRepository;

import java.util.UUID;

@Service
public class StudentServiceImpl implements PersonService<StudentDto> {
    private StudentMapper mapper;
    private StudentRepository repository;

    @Autowired
    public StudentServiceImpl(StudentMapper mapper, StudentRepository repository) {
        this.mapper = mapper;
        this.repository = repository;
    }

    @Override
    public StudentDto findById(UUID id) {
        return mapper.toDto(repository.findById(id).orElseThrow(() -> new StudentNotFoundException(id)));
    }

    @Override
    public Page<StudentDto> findAll(Pageable pageable) {
        return repository.findAll(pageable).map(mapper::toDto);
    }

    @Override
    public StudentDto save(StudentDto student) {
        return mapper.toDto(repository.save(mapper.toEntity(student)));
    }

    @Override
    public void delete(UUID id) {
        repository.delete(repository.findById(id).get());
    }

    @Override
    public Boolean personExists(StudentDto personDto) {
        return repository.existsStudentByFirstNameAndLastNameAndBirthDay(personDto.getFirstName(),
                personDto.getLastName(), personDto.getBirthDay());
    }


}

File: university-service/src/main/java/ua/com/foxminded/courseproject/service/StudentServiceImpl.kt
package ua.com.foxminded.courseproject.service

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import ua.com.foxminded.courseproject.dto.StudentDto
import ua.com.foxminded.courseproject.exceptions.StudentConflictException
import ua.com.foxminded.courseproject.exceptions.StudentNotFoundException
import ua.com.foxminded.courseproject.mapper.StudentMapper
import ua.com.foxminded.courseproject.repository.StudentRepositoryImp
import java.util.*

@Service
class StudentServiceImpl @Autowired constructor(
    private val mapper: StudentMapper,
    private val repository: StudentRepositoryImp
) : PersonService<StudentDto> {
    override fun findById(id: UUID): Mono<StudentDto> {
        return repository.findById(id)
            .switchIfEmpty(Mono.error(StudentNotFoundException(id)))
            .map { mapper.toDto(it) }
    }

    override fun findAll(pageable: Pageable): Flux<StudentDto> {
        return repository.findAll().map { mapper.toDto(it) }
    }

    override fun save(student: StudentDto): Mono<StudentDto> {
//        if (personExists(student).block() == true) {
//            throw StudentConflictException(student)
//        }
        return personExists(student).flatMap {
            if (it == true) {
                Mono.error(StudentConflictException(student))
            } else {
                repository.save(mapper.toEntity(student)!!).map { mapper.toDto(it) }
            }
        }


    }

    override fun delete(id: UUID) {
        repository.findById(id).subscribe { repository.delete(it).subscribe() }
    }

    override fun personExists(personDto: StudentDto): Mono<Boolean> {
        return repository.existsStudentByFirstNameAndLastNameAndBirthDay(
            personDto.firstName!!,
            personDto.lastName!!, personDto.birthDay!!
        )
    }
}
